translate/storage/lisa.py

```diff
@@ -329,20 +334,45 @@ def serialize_hook(treestring):
     def serialize(self, out=None):
         """Converts to a string containing the file's XML"""
         root = self.document.getroot()
+        xml_quote_format = "'"
+        xml_encoding = self.encoding.lower()
+
         if self.XMLdoublequotes:
-            if self.XMLuppercaseEncoding:
-                out.write(b'<?xml version="1.0" encoding="UTF-8"?>\n')
-            else:
-                out.write(b'<?xml version="1.0" encoding="utf-8"?>\n')
+            xml_quote_format = '"'
+
+        if self.XMLuppercaseEncoding:
+            xml_encoding = self.encoding.upper()
+
+        xml_declaration = f"<?xml version={xml_quote_format}1.0{xml_quote_format} encoding={xml_quote_format}{xml_encoding}{xml_quote_format}?>\n"
+
+        out.write(xml_declaration.encode(self.encoding))
+
         if self.XMLindent:
             reindent(root, **self.XMLindent)
-        treestring = etree.tostring(
-            self.document,
-            pretty_print=not self.XMLindent,
-            xml_declaration=not self.XMLdoublequotes,
-            encoding=self.encoding,
-            doctype=self.XMLdoctype,
-        )
+        # When XMLSelfClosingTags flag is active, relay
+        # against LXML library - otherwise we're using
+        # standard xml library.
+        if self.XMLSelfClosingTags:
+            treestring = etree.tostring(
+                self.document,
+                pretty_print=not self.XMLindent,
+                xml_declaration=False,
+                encoding=self.encoding,
+                doctype=self.XMLdoctype,
+            )
+        else:
+            if self.XMLdoctype:
+                out.write(f"{self.XMLdoctype}\n".encode(self.encoding))
+            stream = io.BytesIO()
+
+            ElementTree.register_namespace("", self.namespace)
+            ElementTree.ElementTree(root).write(
+                stream,
+                self.encoding,
+                short_empty_elements=False,
+            )
+            treestring = stream.getvalue()
+
```

Comment:
Okay, I think we need to revisit this. `ElementTree.register_namespace` will definitely have tendency to break behavior in some situations (at least in any async or threaded processes). We should stick to using parameters only, not altering the global state.

According to https://stackoverflow.com/a/34177354/225718 `method="c14n"` should do the trick for lxml.

So, how about this?

```suggestion
        if self.XMLSelfClosingTags and self.encoding == "utf-8":
             kwargs = {"method": "c14n"}
        else:
             kwargs = {"encoding": self.encoding}

        treestring = etree.tostring(
            self.document,
            pretty_print=not self.XMLindent,
            xml_declaration=False,
            doctype=self.XMLdoctype,
            **kwargs,
        )
```

Reply:
I think it does
I'll make further check tomorrow afternoon